certbot-ci/launch-integration-tests.py

```diff
@@ -0,0 +1,232 @@
+
+        if covered < COVERAGE_THRESHOLD:
+            sys.stderr.write('Current coverage ({0}) is under threshold ({1})!{2}'
+                             .format(round(covered, 2),
+                                     COVERAGE_THRESHOLD, os.linesep))
+
+            return 1
+
+    return 0
+
+
+def main(cli_args=sys.argv[1:]):
+    # type: (List[str]) -> int
+    """
+    Main function of the CLI tool.
+    Will:
+        - expose the argument parser
+        - process user input
+        - check runtime compatiblity (Docker, docker-compose, Nginx)
+        - generate the pytest command
+        - create a temporary workspace and the persistent GIT repositories space
+        - execute the pytest command in the temporary workspace
+        - launch the coverage if needed
+    :param cli_args: The arguments given by the CLI
+    :return: 0 if the integration tests execution is considered as successful, 1 otherwise
+    :rtype: int
+    """
+    main_parser = create_parser()
+    args = main_parser.parse_args(cli_args)
+
+    # Check for runtime compatiblity: some tools are required to be available in PATH
+    try:
+        subprocess.check_output(['docker', '-v'], stderr=subprocess.STDOUT)
+    except (subprocess.CalledProcessError, OSError):
+        raise ValueError('Error: docker is required in PATH to launch the integration tests, '
+                         'but is not installed or not available for current user.')
+
+    try:
+        subprocess.check_output(['docker-compose', '-v'], stderr=subprocess.STDOUT)
+    except (subprocess.CalledProcessError, OSError):
+        raise ValueError('Error: docker-compose is required in PATH to launch the integration tests, '
+                         'but is not installed or not available for current user.')
+
+    # We will know, after the next command, how many workers need to be setup. Pytest command
+    # will contain options that rely on several pytest plugins. Theses plugins should have been
+    # installed through pip using the setup.py of certbot-ci project.
+    (command, workers) = prepare_pytest_command(args)
+
+    repositories_path = os.path.join(CURRENT_DIR, '.ci_assets/integration_tests')
+    try:
+        os.makedirs(repositories_path)
+    except OSError as error:
+        if error.errno != errno.EEXIST:
+            raise
+
+    # By calling certbot_ci_workspace, the tempfile.tempdir value of standard tempfile module is
+    # modified for the context execution time, to ensure that any temporary assets will be created
+    # under the cerbot-ci workspace.
+    with certbot_ci_workspace():
+        acme_config = {}
+        # Prepare the acme config server. Data is specific to an acme type. Module
+        # utils.acme_server will handle theses specifics.
+        if 'pebble' in args.acme_server:
+            acme_config['type'] = 'pebble'
+            acme_config['option'] = 'nonstrict' if 'nonstrict' in args.acme_server else 'strict'
```

Comment:
It is two different runtime modes for Pebble. Because Pebble is quite on the edge of ACME protocol specifications, it can be difficult to use it as an integration tool to build software that will use real ACME CA servers. Because theses servers have options that are removed from Pebble in `strict` mode (deprecated options, security problem and so on.).

So the `nonstrict` mode is here to allow some reasonnable deprecated options (things that are not deprecated because of a security flaw) to be used in Pebble. It should be the standard mode for Certbot, as it mimic almost all Boulder ACME v2 capabilities, apart obviously TLS-SNI-01 challenges.